lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java

```diff
@@ -2262,4 +2341,539 @@ public void onError(ErrorInfo reason) {
+
+			try {
+				PresenceMessage[] presenceMessages = channel.presence.get(testClientId1, false);
+				assertEquals("Verify total number of received presence messages", presenceMessages.length, 1);
+				assertEquals("Verify present message is valid", presenceMessages[0].data, presenceData);
+			} catch (InterruptedException e) {}
+		} finally {
+			if (ably != null)
+				ably.close();
+		}
+	}
+
+	/**
+	 * Test if Presence.syncComplete() works. Enter client on one ably connection and
+	 * wait for initial SYNC on another connection.
+	 *
+	 * Tests RTP13
+	 */
+	@Test
+	public void sync_complete() {
+		AblyRealtime ably1 = null, ably2 = null;
+		final String channelName = "sync_complete" + testParams.name;
+		try {
+			ClientOptions opts = createOptions(testVars.keys[0].keyStr);
+			ably1 = new AblyRealtime(opts);
+			ably2 = new AblyRealtime(opts);
+
+			Channel channel1 = ably1.channels.get(channelName);
+			channel1.presence.enterClient(testClientId1);
+			new PresenceWaiter(Action.enter, channel1).waitFor(1);
+
+			Channel channel2 = ably2.channels.get(channelName);
+			assertFalse("Verify SYNC is not complete yet", channel2.presence.syncComplete());
+			channel2.attach();
+			/* Wait for the SYNC to complete */
+			new PresenceWaiter(Action.present, channel2).waitFor(1);
+			try {
+				channel2.presence.get(true);
+			} catch (InterruptedException e) {}
+			/* Initial SYNC should be complete at this point */
+			assertTrue("Verify SYNC is complete", channel2.presence.syncComplete());
+		} catch (AblyException e) {
+			fail("Unexpected exception");
+		} finally {
+			if (ably1 != null)
+				ably1.close();
+			if (ably2 != null)
+				ably2.close();
+		}
+	}
+
+	/**
+	 * Enter client without permission to do so, check exception
+	 * Tests RTP8h
+	 */
+	@Test
+	public void presence_enter_without_permission() throws AblyException {
+		String channelName = "presence_enter_without_permission" + testParams.name;
+		AblyRealtime ably = null;
+
+		try {
+			/* init ably for token */
+			ClientOptions optsForToken = createOptions(testVars.keys[0].keyStr);
+			final AblyRest ablyForToken = new AblyRest(optsForToken);
+
+			/* get first token */
+			Auth.TokenParams tokenParams = new Auth.TokenParams();
+			Capability capability = new Capability();
+			capability.addResource(channelName, "publish");	/* no presence permission! */
+			tokenParams.capability = capability.toString();
+			tokenParams.clientId = testClientId1;
+
+			Auth.TokenDetails token = ablyForToken.auth.requestToken(tokenParams, null);
+			assertNotNull("Expected token value", token.token);
+
+			ClientOptions opts = createOptions();
+			opts.clientId = testClientId1;
+			opts.tokenDetails = token;
+			ably = new AblyRealtime(opts);
+
+			Channel channel = ably.channels.get(channelName);
+			channel.attach();
+
+			ChannelWaiter channelWaiter = new ChannelWaiter(channel);
+			channelWaiter.waitFor(ChannelState.attached);
+
+			final boolean[] flags = new boolean[] {
```

Comment:
I don't propose you change this, but I don't follow why you would use cryptic flags over simply two varaibles `boolean callbackCalled` and `boolean success`?